packages/captp/src/captp.js

```diff
@@ -0,0 +1,703 @@
+// @ts-check
+
+// Your app may need to `import '@agoric/eventual-send/shim'` to get HandledPromise
+
+// This logic was mostly lifted from @agoric/swingset-vat liveSlots.js
+// Defects in it are mfig's fault.
+import { Remotable, Far, makeMarshal, QCLASS } from '@agoric/marshal';
+import { E, HandledPromise } from '@agoric/eventual-send';
+import { isPromise, makePromiseKit } from '@agoric/promise-kit';
+import { assert, details as X } from '@agoric/assert';
+
+import { makeTrap } from './trap.js';
+
+import './types.js';
+
+export { E };
+
+/**
+ * @param {any} maybeThenable
+ * @returns {boolean}
+ */
+const isThenable = maybeThenable =>
+  maybeThenable && typeof maybeThenable.then === 'function';
+
+/**
+ * @typedef {Object} CapTPOptions the options to makeCapTP
+ * @property {(err: any) => void} onReject
+ * @property {number} epoch an integer tag to attach to all messages in order to
+ * assist in ignoring earlier defunct instance's messages
+ * @property {TrapGuest} trapGuest if specified, enable this CapTP (guest) to
+ * use Trap(target) to block while the recipient (host) resolves and
+ * communicates the response to the message
+ * @property {TrapHost} trapHost if specified, enable this CapTP (host) to serve
+ * objects marked with makeTrapHandler to synchronous clients (guests)
+ */
+
+/**
+ * Create a CapTP connection.
+ *
+ * @param {string} ourId our name for the current side
+ * @param {(obj: Record<string, any>) => void} rawSend send a JSONable packet
+ * @param {any} bootstrapObj the object to export to the other side
+ * @param {Partial<CapTPOptions>} opts options to the connection
+ */
+export const makeCapTP = (
```

Comment:
Would it always be incorrect to have both a `trapGuest` and a `trapHost`? If so, should we assert that we do not have both?

Reply:
That's a good idea.  Will do.